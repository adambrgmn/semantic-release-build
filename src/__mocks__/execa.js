/* eslint-disable no-underscore-dangle */

class ExecaError extends Error {
  constructor(msg, { stdout, stderr, code, killed, signal, cmd, timedOut }) {
    super(msg);
    this.stdout = stdout;
    this.stderr = stderr;
    this.code = code;
    this.killed = killed;
    this.signal = signal;
    this.cmd = cmd;
    this.timedOut = timedOut;
  }
}

const execa = jest.fn((cmd, args) => {
  const ret = {
    stdout: 'Success',
    stderr: '',
    code: 0,
    failed: false,
    killed: false,
    signal: null,
    cmd: `${cmd} ${args.join(' ')}`,
    timedOut: false,
  };

  if (execa.__shouldThrow) {
    const msg = `${ret.cmd} failed`;
    const err = new ExecaError(msg, {
      ...ret,
      stdout: '',
      stderr: msg,
      code: 1,
      failed: true,
    });

    return Promise.reject(err);
  }

  return Promise.resolve(ret);
});

execa.__shouldThrow = false;
execa.__setShouldThrow = b => {
  execa.__shouldThrow = b;
};

module.exports = execa;
