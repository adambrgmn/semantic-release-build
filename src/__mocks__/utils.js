/* eslint-disable no-underscore-dangle */

let __packageManager = 'npm';
const __setPackageManager = pm => {
  __packageManager = pm;
};

const getPackageManager = async () => __packageManager;

const getPkgData = async () => ({ path: '/package.json' });

module.exports = {
  getPkgData,
  getPackageManager,
  __setPackageManager,
};
