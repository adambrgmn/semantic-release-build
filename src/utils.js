const { promisify } = require('util');
const path = require('path');
const fs = require('fs');
const which = promisify(require('which'));
const readPkgUp = require('read-pkg-up');

const access = promisify(fs.access);
const realpath = promisify(fs.realpath);

const hasFile = file =>
  realpath(process.cwd(), null)
    .then(cwd => readPkgUp({ cwd }))
    .then(({ path: pkgPath }) => {
      const projectDir = path.dirname(pkgPath);
      const fromRoot = (...p) => path.join(projectDir, ...p);

      return access(fromRoot(file), fs.constants.F_OK).then(
        () => true,
        () => false,
      );
    });

const getPackageManager = () =>
  Promise.all([which('yarn'), hasFile('yarn.lock')])
    .then(([hasBin, hasLockFile]) => (hasBin && hasLockFile ? 'yarn' : 'npm'))
    .catch(() => 'npm');

module.exports = {
  hasFile,
  getPackageManager,
};
