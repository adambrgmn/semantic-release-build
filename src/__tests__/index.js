/* eslint-disable no-underscore-dangle */
import execa from 'execa';
import * as utils from '../utils';
import { publish } from '../index';

jest.mock('fs');
jest.mock('execa');
jest.mock('../utils');

const logger = { log: jest.fn() };
const getReleaseConfig = () => ({ nextRelease: { version: '1.0.0' }, logger });

beforeEach(() => {
  jest.resetAllMocks();
});

test('Calls execa', async () => {
  await publish({}, getReleaseConfig());
  expect(execa.mock.calls).toMatchSnapshot();
});

test('Accepts flags as config', async () => {
  await publish({ flags: '--out-dir build' }, getReleaseConfig());
  expect(execa.mock.calls).toMatchSnapshot();
});

test('Skips extra -- when packagemanger is yarn', async () => {
  utils.__setPackageManager('yarn');
  await publish({ flags: '--out-dir build' }, getReleaseConfig());
  expect(execa.mock.calls).toMatchSnapshot();
});

test('Throws if options.flags is set but not as string', async () => {
  await expect(publish({ flags: 123 }, getReleaseConfig())).rejects.toThrow();
});

test('Throws if spawn return non-0 status', async () => {
  execa.__setShouldThrow(true);
  await expect(publish()).rejects.toThrow();
});
