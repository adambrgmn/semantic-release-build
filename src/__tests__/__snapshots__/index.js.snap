// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accepts flags as config 1`] = `
Array [
  Array [
    "npm",
    Array [
      "run",
      "build",
      "--",
      "--out-dir",
      "build",
    ],
  ],
]
`;

exports[`Calls execa 1`] = `
Array [
  Array [
    "npm",
    Array [
      "run",
      "build",
    ],
  ],
]
`;

exports[`Skips extra -- when packagemanger is yarn 1`] = `
Array [
  Array [
    "yarn",
    Array [
      "run",
      "build",
      "--out-dir",
      "build",
    ],
  ],
]
`;
